plasmapy/utils/decorators/checks.py

```diff
@@ -1,20 +1,1083 @@
 """
 Decorator for checking input/output arguments of functions.
 """
-__all__ = ["check_quantity", "check_relativistic"]
+__all__ = ["check_values", "check_units", "check_quantity", "check_relativistic",
+           "CheckBase", "CheckUnits", "CheckValues"]
 
+import collections
 import functools
 import inspect
 import numpy as np
 import warnings
 
 from astropy import units as u
-from astropy.units import UnitsWarning
 from astropy.constants import c
+from astropy.units import UnitsWarning
+from astropy.units.core import _normalize_equivalencies
+from astropy.units.decorators import _get_allowed_units
+from functools import reduce
 from plasmapy.utils.decorators import preserve_signature
-from plasmapy.utils.exceptions import RelativityWarning, RelativityError
+from plasmapy.utils.exceptions import (PlasmaPyWarning,
+                                       RelativityWarning,
+                                       RelativityError)
 from textwrap import dedent
-from typing import (Dict, Union)
+from typing import (Any, Dict, List, Tuple, Union)
+
+try:
+    from astropy.units.equivalencies import Equivalency
+except ImportError:
+    # astropy defined the Equivalency class in v3.2.1
+    class Equivalency:
+        pass
+
+
+class CheckBase:
+    """
+    Base class for 'Check' decorator classes.
+
+    Parameters
+    ----------
+    checks_on_return
+        specified checks on the return of the wrapped function
+
+    **checks
+        specified checks on the input arguments of the wrapped function
+    """
+    def __init__(self, checks_on_return=None, **checks):
+        self._checks = checks
+        if checks_on_return is not None:
+            self._checks['checks_on_return'] = checks_on_return
+
+    @property
+    def checks(self):
+        """
+        Requested checks on the decorated function's input arguments
+        and/or return.
+        """
+        return self._checks
```

Comment:
Yeah, I've always liked properties for key variables, since (1) it's easier to produce clean documentation, (2) can make code more readable IMHO, and (3) it gives me some setter and getter controls down the line if needed.